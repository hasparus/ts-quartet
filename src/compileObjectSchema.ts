import { addTabs } from "./addTabs";
import { getKeyAccessor } from "./getKeyAccessor";
import { handleSchema } from "./handleSchema";
import { toContext } from "./toContext";
import { CompilationResult, IObjectSchema, Prepare, Schema } from "./types";

function compilePropValidationWithoutRest(
  c: (schema: Schema) => CompilationResult,
  key: string,
  valueId: string,
  ctxId: string,
  schema: Schema,
  preparations: Prepare[],
  stringNumbersSymbols: Array<string | number | symbol>
): string {
  return handleSchema({
    constant: constant => {
      if (constant === null) {
        return `if (${valueId} !== null) return false`;
      }
      if (constant === undefined) {
        return `if (${valueId} !== undefined) return false`;
      }
      if (typeof constant === "number") {
        if (Number.isNaN(constant)) {
          return `if (!Number.isNaN(${valueId})) return false`;
        }
        return `if (${valueId} !== ${constant}) return false`;
      }
      if (constant === "true" || constant === "false") {
        return `if (${valueId} !== '${constant}') return false`;
      }
      if (typeof constant === "symbol" || typeof constant === "string") {
        stringNumbersSymbols.push(constant);
        return "";
      }
      return `if (${valueId} !== ${JSON.stringify(constant)}) return false`;
    },
    function: funcSchema => {
      const s = funcSchema();
      if (s.prepare) {
        preparations.push(s.prepare);
      }
      const notCheck = s.not
        ? s.not(valueId, ctxId)
        : `!(${s.check(valueId, ctxId)})`;
      return s.handleError
        ? `if (${notCheck}) {\n${addTabs(
            s.handleError(valueId, ctxId)
          )}\n  return false\n}`
        : `if (${notCheck}) return false`;
    },
    object: objectSchema => {
      const keys = Object.keys(objectSchema);
      const codeLines = [`if (!${valueId}) return false`];
      const important: string[] = [];
      // tslint:disable-next-line
      for (let i = 0; i < keys.length; i++) {
        const innerKey = keys[i];
        const innerKeyAccessor = getKeyAccessor(innerKey);
        const keyValidValues: Array<string | symbol> = [];
        const innerKeyId = valueId + innerKeyAccessor;
        const code = compilePropValidationWithoutRest(
          c,
          innerKey,
          innerKeyId,
          ctxId,
          objectSchema[innerKey],
          preparations,
          keyValidValues
        );

        if (keyValidValues.length > 0) {
          for (const valid of keyValidValues) {
            const [keyConstantId, prepare] = toContext(innerKeyId, valid);
            preparations.push(prepare);
            important.push(
              `if (${innerKeyId} !== ${ctxId}['${keyConstantId}']) return false`
            );
          }
        }
        if (code) {
          codeLines.push(code);
        }
      }
      codeLines.splice(1, 0, ...important);
      return codeLines.join("\n");
    },
    objectRest: objectSchema => {
      const compiled = c(objectSchema);
      const [id, prepare] = toContext(key, compiled);
      preparations.push(prepare);
      return compilePropValidationWithoutRest(
        c,
        key,
        valueId,
        ctxId,
        () => ({
          check: () => `${ctxId}['${id}'](${valueId})`,
          handleError: () =>
            `${ctxId}.explanations.push(...${ctxId}['${id}'].explanations)`,
          not: () => `!${ctxId}['${id}'](${valueId})`
        }),
        preparations,
        stringNumbersSymbols
      );
    },
    variant: schemas => {
      if (schemas.length === 0) {
        return `return false`;
      }
      if (schemas.length === 1) {
        return compilePropValidationWithoutRest(
          c,
          key,
          valueId,
          ctxId,
          schemas[0],
          preparations,
          stringNumbersSymbols
        );
      }
      const compiled = c(schemas);
      const [id, prepare] = toContext(key, compiled);
      preparations.push(prepare);
      return compilePropValidationWithoutRest(
        c,
        key,
        valueId,
        ctxId,
        () => ({
          check: () => `${ctxId}['${id}'](${valueId})`,
          handleError: () =>
            `${ctxId}.explanations.push(...${ctxId}['${id}'].explanations)`,
          not: () => `!${ctxId}['${id}'](${valueId})`
        }),
        preparations,
        stringNumbersSymbols
      );
    }
  })(schema);
}

export function compileObjectSchema(
  c: (schema: Schema) => CompilationResult,
  s: IObjectSchema
) {
  const keys = Object.keys(s);
  if (keys.length === 0) {
    return Object.assign((v: any) => !!v, { explanations: [] });
  }
  const bodyCodeLines = [];
  const preparations: Prepare[] = [];
  const ctxId = "validator";
  const validValues: Record<string, Record<string, true>> = {};
  const withValidValues: string[] = [];
  // tslint:disable-next-line
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    const schema = s[key];
    const keyAccessor = getKeyAccessor(key);
    const valueId = `value${keyAccessor}`;
    const keyValidValues: any[] = [];
    const codeLine = compilePropValidationWithoutRest(
      c,
      key,
      valueId,
      ctxId,
      schema,
      preparations,
      keyValidValues
    ).trim();
    if (codeLine) {
      bodyCodeLines.push(codeLine);
    }
    if (keyValidValues.length > 0) {
      withValidValues.push(key);
      if (!validValues[key]) {
        validValues[key] = {};
      }
      for (const valid of keyValidValues) {
        validValues[key][valid] = true;
      }
    }
  }

  if (withValidValues.length > 0) {
    preparations.push(context => {
      context.__validValues = validValues;
    });
    bodyCodeLines.unshift(
      ...withValidValues.map(key => {
        const keyAccessor = getKeyAccessor(key);
        return `if (validator.__validValues${keyAccessor}[value${keyAccessor}] !== true) return false`;
      })
    );
  }
  bodyCodeLines.unshift("if (!value) return false");
  if (bodyCodeLines.some(line => line.indexOf("explanations") >= 0)) {
    bodyCodeLines.unshift("validator.explanations = []");
  }
  const code = `
    (() => {\nfunction validator(value) {\n${addTabs(
      bodyCodeLines.join("\n")
    )}\n  return true\n}
      return validator
    })()
  `.trim();

  // tslint:disable-next-line
  const ctx = eval(code);

  for (const prepare of preparations) {
    prepare(ctx);
  }

  ctx.explanations = [];

  return ctx;
}
