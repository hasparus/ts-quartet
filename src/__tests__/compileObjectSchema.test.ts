import { compileObjectSchema } from "../compileObjectSchema";
import { v } from "../index";
import { getDescription } from "./getDescription";

describe("compileObjectSchema", () => {
  test("empty obj", () => {
    const validator = compileObjectSchema(v, {});
    expect(getDescription(validator)).toMatchInlineSnapshot(`
                                                      Object {
                                                        "_": "function (value) { return value; }",
                                                        "explanations": Array [],
                                                      }
                                    `);
  });
  test("obj with func", () => {
    const validator = compileObjectSchema(v, {
      id: () => ({
        check: id => `${id} === 42`,
        not: id => `${id} !== 42`,
        handleError: (id, ctx) => `${ctx}.explanations.push(${id})`
      })
    });
    expect(validator({})).toBe(false);
    expect(validator.explanations).toEqual([undefined]);
    expect(validator(null)).toBe(false);
    expect(validator.explanations).toEqual([]);
    expect(validator({ id: 43 })).toBe(false);
    expect(validator.explanations).toEqual([43]);

    expect(getDescription(validator)).toMatchInlineSnapshot(`
                                                Object {
                                                  "_": "function validator(value) {
                                                    validator.explanations = []
                                                    if (!value) return false
                                                    if (value.id !== 42) {
                                                      validator.explanations.push(value.id)
                                                      return false
                                                    }
                                                    return true
                                                  }",
                                                  "__validValues": Object {},
                                                  "explanations": Array [
                                                    43,
                                                  ],
                                                }
                                `);
  });
  test("obj with constant", () => {
    const validator = compileObjectSchema(v, {
      id: 42,
      nan: NaN,
      inf: Infinity,
      minf: -Infinity,
      zero: -0,
      str: "Andrew",
      f: false,
      t: true,
      symb: Symbol.for("test")
    });
    expect(validator({})).toBe(false);
    expect(validator.explanations).toEqual([]);
    expect(validator(null)).toBe(false);
    expect(validator.explanations).toEqual([]);
    expect(validator({ id: 43 })).toBe(false);
    expect(validator.explanations).toEqual([]);
    expect(
      validator({
        id: 42,
        nan: NaN,
        inf: Infinity,
        minf: -Infinity,
        zero: 0,
        str: "Andrew",
        f: false,
        t: true,
        symb: Symbol.for("test")
      })
    ).toBe(true);
    expect(validator.explanations).toEqual([]);
    expect(validator({ id: "42" })).toBe(false);
    expect(validator.explanations).toEqual([]);

    expect(getDescription(validator)).toMatchInlineSnapshot(`
                        Object {
                          "_": "function validator(value) {
                            validator.explanations = []
                            if (!value) return false
                            if (!validator.__validValues.str[value.str]) return false
                            if (!validator.__validValues.symb[value.symb]) return false
                            if (value.id !== 42) return false
                            if (!Number.isNaN(value.nan)) return false
                            if (value.inf !== Infinity) return false
                            if (value.minf !== -Infinity) return false
                            if (value.zero !== 0) return false
                            if (value.f !== false) return false
                            if (value.t !== true) return false
                            return true
                          }",
                          "__validValues": Object {
                            "str": Object {
                              "Andrew": true,
                            },
                            "symb": Object {
                              Symbol(test): true,
                            },
                          },
                          "explanations": Array [],
                        }
                `);
  });
  test("obj with variants", () => {
    const validator = compileObjectSchema(v, {
      gender: ["male", "female"],
      grade: Array.from({ length: 11 }, (_, i) => i + 1)
    });

    expect(validator({})).toBe(false);
    expect(validator({ gender: "male" })).toBe(false);
    expect(validator({ grade: 10 })).toBe(false);
    expect(validator({ grade: "10" })).toBe(false);
    expect(validator({ gender: "male", grade: 0 })).toBe(false);
    expect(validator({ grade: 12, gender: "female" })).toBe(false);
    expect(validator({ grade: "10" })).toBe(false);
    expect(validator({ grade: "10", gender: "male" })).toBe(false);
    expect(validator({ grade: 10, gender: "male" })).toBe(true);

    expect(getDescription(validator)).toMatchInlineSnapshot(`
            Object {
              "_": "function validator(value) {
                validator.explanations = []
                if (!value) return false
                if (!validator['gender-0'](value.gender)) return false
                if (!validator['grade-0'](value.grade)) return false
                return true
              }",
              "__validValues": Object {},
              "explanations": Array [],
              "gender-0": "function validator(value) {
                if (validator.__validValuesDict[value] === true) return true
                return false
              }",
              "gender-0.__validValuesDict": Object {
                "female": true,
                "male": true,
              },
              "gender-0.explanations": Array [],
              "grade-0": "function validator(value) {
                if (value === 1) return true
                if (value === 2) return true
                if (value === 3) return true
                if (value === 4) return true
                if (value === 5) return true
                if (value === 6) return true
                if (value === 7) return true
                if (value === 8) return true
                if (value === 9) return true
                if (value === 10) return true
                if (value === 11) return true
                return false
              }",
              "grade-0.explanations": Array [],
            }
        `);
  });
  test("obj + obj", () => {
    const validator = compileObjectSchema(v, {
      deep: { deep: { space: "true" } }
    });
    expect(validator({ deep: { deep: { space: "false" } } })).toBe(false);
    expect(validator({ deep: { deep: { space: "true" } } })).toBe(true);
    expect(getDescription(validator)).toMatchInlineSnapshot(`
      Object {
        "_": "function validator(value) {
          validator.explanations = []
          if (!value) return false
          if (!validator['deep-1'](value.deep)) return false
          return true
        }",
        "__validValues": Object {},
        "deep-1": "function validator(value) {
          validator.explanations = []
          if (!value) return false
          if (!validator['deep-0'](value.deep)) return false
          return true
        }",
        "deep-1.__validValues": Object {},
        "deep-1.deep-0": "function validator(value) {
          validator.explanations = []
          if (!value) return false
          if (value.space !== 'true') return false
          return true
        }",
        "deep-1.deep-0.__validValues": Object {},
        "deep-1.deep-0.explanations": Array [],
        "deep-1.explanations": Array [],
        "explanations": Array [],
      }
    `);
  });
});
