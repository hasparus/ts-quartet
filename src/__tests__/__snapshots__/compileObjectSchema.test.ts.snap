// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileObjectSchema Obj: toString string constant value 1`] = `
Object {
  "_": "function validator(value) {
    if (!value) return false
    if (validator.__validValues.a[value.a] !== true) return false
    return true
  }",
  "__validValues": Object {
    "a": Object {
      "A": true,
    },
  },
  "explanations": Array [],
}
`;

exports[`compileObjectSchema empty obj 1`] = `
Object {
  "_": "function (value) { return value; }",
  "explanations": Array [],
}
`;

exports[`compileObjectSchema obj + obj 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!value.deep) return false
    if (value.deep.another !== validator['value.deep.another-0']) return false
    if (!value.deep.deep) return false
    if (value.deep.deep.A !== validator['value.deep.deep.A-0']) return false
    if (!(value.deep.deep.space === 'true')) {
      validator.explanations.push(value.deep.deep.space)
      return false
    }
    return true
  }",
  "explanations": Array [],
  "value.deep.another-0": "B",
  "value.deep.deep.A-0": "A",
}
`;

exports[`compileObjectSchema obj with constant 1`] = `
Object {
  "_": "function validator(value) {
    if (!value) return false
    if (validator.__validValues.str[value.str] !== true) return false
    if (validator.__validValues.symb[value.symb] !== true) return false
    if (value.id !== 42) return false
    if (!Number.isNaN(value.nan)) return false
    if (value.inf !== Infinity) return false
    if (value.minf !== -Infinity) return false
    if (value.zero !== 0) return false
    if (value.f !== false) return false
    if (value.t !== true) return false
    return true
  }",
  "__validValues": Object {
    "str": Object {
      "Andrew": true,
    },
    "symb": Object {
      Symbol(test): true,
    },
  },
  "explanations": Array [],
}
`;

exports[`compileObjectSchema obj with func 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (value.id !== 42) {
      validator.explanations.push(value.id)
      return false
    }
    return true
  }",
  "explanations": Array [
    43,
  ],
}
`;

exports[`compileObjectSchema obj with variants 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!validator['gender-0'](value.gender)) {
      validator.explanations.push(...validator['gender-0'].explanations)
      return false
    }
    if (!validator['grade-0'](value.grade)) {
      validator.explanations.push(...validator['grade-0'].explanations)
      return false
    }
    return true
  }",
  "explanations": Array [],
  "gender-0": "function validator(value) {
    if (validator.__validValuesDict[value] === true) return true
    return false
  }",
  "gender-0.__validValuesDict": Object {
    "female": true,
    "male": true,
  },
  "gender-0.explanations": Array [],
  "grade-0": "function validator(value) {
    if (value === 1) return true
    if (value === 2) return true
    if (value === 3) return true
    if (value === 4) return true
    if (value === 5) return true
    if (value === 6) return true
    if (value === 7) return true
    if (value === 8) return true
    if (value === 9) return true
    if (value === 10) return true
    if (value === 11) return true
    return false
  }",
  "grade-0.explanations": Array [],
}
`;

exports[`compileObjectSchema obj: variant explanations 3`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!validator['var-0'](value.var)) {
      validator.explanations.push(...validator['var-0'].explanations)
      return false
    }
    return true
  }",
  "explanations": Array [],
  "var-0": "function validator(value) {
    validator.explanations = []
    if (value === \\"A\\") return true;
    if (value === \\"B\\") return true;
    validator.explanations.push(\\"Is not A\\")
    validator.explanations.push(\\"Is not B\\")
    return false
  }",
  "var-0.explanations": Array [],
}
`;
